core/scale/scale_decoder_stream.cpp

```diff
@@ -5,4 +5,149 @@
 
 #include "scale/scale_decoder_stream.hpp"
 
-namespace kagome::scale {}
+#include "macro/unreachable.hpp"
+#include "scale/outcome_throw.hpp"
+#include "scale/scale_error.hpp"
+#include "scale/types.hpp"
+
+namespace kagome::scale {
+  namespace {
+    CompactInteger decodeCompactInteger(ScaleDecoderStream &stream) {
+      auto first_byte = stream.nextByte();
+
+      const uint8_t flag = (first_byte)&0b00000011u;
+
+      size_t number = 0u;
+
+      switch (flag) {
+        case 0b00u: {
+          number = static_cast<size_t>(first_byte >> 2u);
+          break;
+        }
+
+        case 0b01u: {
+          auto second_byte = stream.nextByte();
+
+          number = (static_cast<size_t>((first_byte)&0b11111100u)
+                    + static_cast<size_t>(second_byte) * 256u)
+              >> 2u;
+          break;
+        }
+
+        case 0b10u: {
+          number = first_byte;
+          size_t multiplier = 256u;
+          if (!stream.hasMore(3u)) {
+            // not enough data to decode integer
+            common::raise(DecodeError::NOT_ENOUGH_DATA);
+            UNREACHABLE
+          }
+
+          for (auto i = 0u; i < 3u; ++i) {
+            // we assured that there are 3 more bytes,
+            // no need to make checks in a loop
+            number += (stream.nextByte()) * multiplier;
+            multiplier = multiplier << 8u;
+          }
+          number = number >> 2u;
+          break;
+        }
+
+        case 0b11: {
+          auto bytes_count = ((first_byte) >> 2u) + 4u;
+          if (!stream.hasMore(bytes_count)) {
+            // not enough data to decode integer
+            common::raise(DecodeError::NOT_ENOUGH_DATA);
+            UNREACHABLE
```

Comment:
isn't it easier to add UNREACHABLE inside raise?

Reply:
it doesn't work like this. we need to place UNREACHABLE exactly in the function where "unreachable" paths exist, but not deduced by compiler.

Reply:
Refactored. Actually we don't need it in raise function. Though, it isn't required at the place you mentioned.